### ISSUE TITLE
Possible Bug in StackScreenNavigator popping to the same Screen over and over.

### ISSUE THREAD

**Issue by GV3** [BowlerHatLLC/feathers#1386]
This is what I'm experiencing:

- Being in Screen A
- push Screen B via the navigator
- then in B (with content b1 passed via settings) a screen C is pushed with the navigator
- Then screen C is replaced with screen B (with content b2 passed via settings). So basically is the same Screen instance although it does look different.
- Then screen B should be popped via COMPLETE signal and go to A... but it goes back to itself instead, to the same instance of B,
- The second time it goes to A

(Note that if we go multiple times from B to C and then to B, is will go the same amount of times from  B to B until it can go to A)

If I were passing a Class in the creation of the StackScreenNavigator I could maybe understand, but all of this screen items are using unique instances of the Screens.

I would expect the Navigator to know that it's the same screen instance and not try to pop to itself.

YES, I can go BACK from C to B but the thing is that in this App you can also go from D to C and then to B... so it's not always required to go back. And YES, right now the code is running IF's to know if it should push or pop, but I think the behavior I expect might have been the original of Feathers and is Buggy

**Comment by joshtynjala** [comment]
This sounds like the correct behavior to me.

When you push C, the stack is [A, B, C]. Then you replace C with B, so the stack is [A, B, B]. The first B is not removed from the stack. You only replaced C. It sounds like you should replace B with C instead of pushing C so that the stack is [A, C].

Alternatively, if A is the root screen when the stack is [A, B, B], you could pop to root and skip the first B.

**Comment by GV3** [comment]
I understand your point, especially in [A, B, B] if the first B is a different instance or ScreenItem than the second B (If we pass a Class when we create a new instance of StackScreenNavigatorItem).

But in the case of [A, B, B] where the first and second B are the exact same screen navigator item, the navigator could be "smart" enough to go back to A when the second B is popped instead of going first to the first B.

**Comment by joshtynjala** [comment]
There's a lot of potential that kind of smartness to break something, in my opinion. I'm going to leave it as-is.

**Comment by GV3** [comment]
Josh, a little detail before you completely forget about this...

Imagine this stack [A, ...., G, D, D] , I noticed that popping D doesn't even transition the D, no animation... not even an Event.CHANGE was dispatched... So I went inside the code and took a look, and noticed that in the method showScreenInternal in BaseScreenNavigator you test if the activeScreen is the same as the one to be popped/shown and exit the method:

```
if(this._activeScreenID == id)
{
    return this._activeScreen;
}
```

Don't dispatch any event, means that the navigator is assuming that the Screen is the same just by the ID.

From this point of view, and since the possibility that two DIFFERENT screens with the same ID does not exist. I would qualify this as a Bug. 

The question is now what is the proper expected behavior:
1 - Assume that two screens with the same ID can be transitioned one to the other and dispatch the proper Events. 
2 - Assume that the screen is the same (as it is right now) and in the popScreen method keep reiterating until the DisplayObject returned by the showScreenInternal is different than the this._activeScreen

I would vote for number TWO :)

**Comment by joshtynjala** [comment]
If the navigator cannot change to another screen with the same ID, that is indeed a bug.

I can't make assumption 2, though. Imagine a UserProfile screen. You could navigate to the profiles of multiple users in the same stack. The screens should be allowed to have the same ID.

**Comment by GV3** [comment]
> I can't make assumption 2, though. Imagine a UserProfile screen. You could navigate to the profiles of multiple users in the same stack. The screens should be allowed to have the same ID.

I agree, screens should be allowed to have the same ID, that is why I previously proposed the extra check to see if the DisplayObject returned by the showScreenInternal is different than the this._activeScreen. 

In the UserProfile example, when you navigate from UserProfile to UserProfile, each of this will be different DisplayObject even if the ID is the same. For this reason, I proposed to test the equality of the this._activeScreen 

Basically, the screenId can be the same, but the screen as a DisplayObject can not.

**Comment by joshtynjala** [comment]
It's still possible for the same display object to show different data at different positions in the stack, though.
